### PR TITLE
fix Issue 21038 - wchar and dchar string alignment should be 2 and 4,…

### PR DESCRIPTION
**Pull request by WalterBright**
… respectively

### CODE REVIEW

File: test/runnable/mars1.d [hunk 1]
```diff
@@ -2124,6 +2124,17 @@ void testMulAssPair()
...
 
+////////////////////////////////////////////////////////////////////////
+// https://issues.dlang.org/show_bug.cgi?id=21038
+
+const(wchar)* x21038 = "xz";
+const(dchar)* name21038 = "abcd";
+
+void test21038()
+{
+    assert((cast(size_t) name21038) % wchar_t.sizeof == 0);
```
Comment (thewilsonator) [inline test/runnable/mars1.d:2135]: what about `dchar`?

**Review by thewilsonator — COMMENTED** [review]
Also this issue is marked major. Either this PR should target stable or the issue should be reprioritised.

**Comment by WalterBright** [comment]
This issue has been there for more than 10 years, and was just reported. So I'm not so sure it's major.

**Comment by Geod24** [comment]
I think the categorization is fair, any bad codegen bug should be major because of how hard they are to track down. Issues which are `rejects-valid` are annoying but trivial to spot. `accepts-invalid` is a notch up, but still at compile time. But bad codegen can be some of the worst bugs to deal with, as you've experienced over the past few days.

**Comment by WalterBright** [comment]
Fixing some bugs can be disruptive, and does one really want regressions in stable?

> But bad codegen can be some of the worst bugs to deal with, as you've experienced over the past few days.

I know all about finding codegen bugs. The original test suite was designed to minimize effort at tracking them down. But the current test suite has evolved into an edifice that makes it perversely difficult to track down problems. For example, the first program built by a new compiler shouldn't be a test harness that requires it to be perfect before it can even run. The compiler's internal bug checks (i.e. asserts) should not be bypassed. The tests themselves should not be randomly failing because of environmental problems. And lastly, the generated logfiles should designed with helping the reader find out where it went wrong.

https://issues.dlang.org/buglist.cgi?bug_status=NEW&bug_status=ASSIGNED&bug_status=REOPENED&keywords=TestSuite%2C%20&keywords_type=allwords&list_id=232627&query_format=advanced

**Comment by WalterBright** [comment]
By the way, I don't even know what "Azure pipelines (Windows x64)" even is. How is it different from the autotester doing the Win64 tests (which do work)? The documentation for the test doesn't say.

**Comment by Geod24** [comment]
> Fixing some bugs can be disruptive, and does one really want regressions in stable?

It can be, but do you think this is the case for this bug ?

> But the current test suite has evolved [...]

All fair points, and no one disagree with you on this. We just need someone to do the job (and some people already stepped up).
To provide context, the evolution you mentioned were made so that we could *improve things*. On the topic of making the logs useful, the output diff feature added by @MoonlightSentinel would have been much harder to do if `tests/run.d` wasn't a D program. And IIRC the reason it switched from using the host compiler (which it originally was) to the built compiler was because DMD was running out of memory while building the code (which is both because the CI machine have limited memory and because DMD's memory consumption have only been going up for the last 5 years) and we needed to use the new `-lowmem` improvement.

**Review by ghost — APPROVED** [review]
~~I volunteeer to rebase it on stable if this is really required.~~
There's a badass conflict in mars1.d.

**Comment by NilsLankila** [comment]
by [seb comment](https://github.com/dlang/dmd/pull/11511#issuecomment-673465847)

File: src/dmd/backend/out.d [hunk 2]
```diff
@@ -204,6 +204,10 @@ version (SCPP)
                 }
                 else
                 {
+                    version (SCPP)
+                        alignOffset(DATA, 2 << dt.DTalign);
+                    version (MARS)
+                        alignOffset(CDATA, 2 << dt.DTalign);
```
Comment (dkorpel) [inline src/dmd/backend/out.d:210]: Should be `1 << dt.DTalign`, or is the alignment deliberately doubled?

**Pull request merged**